_content/posts/dsa-bellman-ford.md

```diff
@@ -0,0 +1,276 @@
+
+
+## Principais aplicações
+
+* **Redes de Computadores:** Utilizado no Routing Information Protocol (RIP) para calcular rotas mais eficientes.
+* **Logística e Transporte:** Planejamento de trajetos e otimização de custos de transporte.
+* **Finanças:** Modelagem de arbitragem e identificação de inconsistências em taxas de câmbio.
+
+
+## Quando usar?
+
+O Bellman-Ford é a escolha certa quando:
+
+* **Pesos negativos estão presentes:** Diferente do Dijkstra, ele funciona corretamente mesmo com pesos negativos.
+* **É necessário detectar ciclos negativos:** Importante em cálculos financeiros e redes de fluxo.
+* **O número de arestas não é muito grande:** O algoritmo tem complexidade O(V × E), então é eficiente quando o grafo não é muito denso.
+Se o grafo não possui pesos negativos e exige uma solução mais rápida, Dijkstra pode ser uma opção melhor.
+
+
+## Bellman-Ford vs. Dijkstra
+
+| **Critério**                     | **Bellman-Ford**                                                | **Dijkstra**                                         |
+|----------------------------------|-----------------------------------------------------------------|------------------------------------------------------|
+| **Pesos negativos**              | ✅ Funciona corretamente                                        | ❌ Não funciona                                      |
+| **Detecção de ciclos negativos** | ✅ Sim                                                          | ❌ Não                                               |
+| **Complexidade**                 | `O(V × E)`                                                      | `O(V log V + E)` (com heap)                          |
+| **Melhor para**                  | Grafos com pesos negativos ou necessidade de detecção de ciclos | Grafos sem pesos negativos, onde rapidez é essencial |
+
+**Conclusão**: Use **Dijkstra** para eficiência quando **não houver pesos negativos**. Escolha **Bellman-Ford** se precisar lidar com **pesos negativos ou detectar ciclos negativos**.
+
+
+## Conceitos fundamentais
+
+Antes de entender o funcionamento do **Algoritmo de Bellman-Ford**, é essencial conhecer alguns conceitos-chave sobre **grafos e caminhos mínimos**.
+
+### Grafos Direcionados e Não-Direcionados
+
+- Um **grafo direcionado** possui **arestas com direção**, ou seja, uma conexão **A → B** não significa que existe **B → A**.
+- Um **grafo não-direcionado** possui conexões **bidirecionais** (A ↔ B).
+- O **Bellman-Ford** pode ser aplicado a ambos, mas em **grafos não-direcionados** com pesos negativos, pode haver **ciclos negativos triviais**, tornando a solução indefinida.
+
+### Pesos nas arestas (Positivos e Negativos)
+
+- Cada aresta em um grafo pode ter um **peso** (ou custo), representando distância, tempo ou custo financeiro.
```

Comment:
nit:
Is it worth mentioning that this is what we call "Grafos Ponderados"?

Reply:
The text was improved to clarify that point